lucille-core/src/main/java/com/kmwllc/lucille/stage/ApplyFileHandlers.java

```diff
@@ -0,0 +1,139 @@
+ *
+ * <br> filePathField (String, Optional): Specify the field in your documents which will has the file path you want to apply handlers to.
+ * Defaults to "file_path". No processing will occur on documents without this field, even if they have the fileContentField present.
+ *
+ * <br> fileContentField (String, Optional): Specify the field in your documents which has the file's contents as an array of bytes.
+ * Defaults to "file_content". When processing a document with a path to a supported file type, the handler will process the array
+ * of bytes found in this field, if present, before having the FileContentFetcher to open an InputStream for the file's contents.
+ *
+ * <br> handlerOptions (Map): Specifies which file types should be handled / processed by this stage. Valid options include:
+ * <br>   csv (Map, Optional): csv config options for handling csv type files. Config will be passed to CSVFileHandler
+ * <br>   json (Map, Optional): json config options for handling json/jsonl type files. Config will be passed to JsonFileHandler
+ *
+ * <br> <b>Note:</b> handlerOptions should contain at least one of the above entries, otherwise, an Exception is thrown.
+ * <br> <b>Note:</b> XML is not supported.
+ *
+ * <br> gcp (Map, Optional): options for handling GoogleCloud files. Include if you are going to process documents with the filePathField set to a Google Cloud URI.
+ * See FileConnector for necessary arguments.
+ *
+ * <br> s3 (Map, Optional): options for handling S3 files. Include if you are going to process documents with the filePathField set to an S3 URI.
+ * See FileConnector for necessary arguments.
+ *
+ * <br> azure (Map, Optional): options for handling Azure files. Include if you are going to process documents with the filePathField set to an Azure URI.
+ * See FileConnector for necessary arguments.
+ *
+ */
+public class ApplyFileHandlers extends Stage {
+
+  private final Config handlerOptions;
+
+  private final String filePathField;
+  private final String fileContentField;
+  private final FileContentFetcher fileFetcher;
+  private final Map<String, FileHandler> fileHandlers;
+
+  public ApplyFileHandlers(Config config) {
+    super(config, new StageSpec()
+        .withOptionalParents("handlerOptions", "gcp", "azure", "s3")
+        .withOptionalProperties("filePathField", "fileContentField"));
+
+    this.handlerOptions = config.getConfig("handlerOptions");
+
+    if (handlerOptions.isEmpty()) {
+      throw new IllegalArgumentException("Must specify at least one file handler.");
+    }
+
+    this.filePathField = ConfigUtils.getOrDefault(config, "filePathField", "file_path");
+    this.fileContentField = ConfigUtils.getOrDefault(config, "fileContentField", "file_content");
+
+    this.fileFetcher = new FileContentFetcher(config);
+    this.fileHandlers = new HashMap<>();
+  }
+
+  @Override
+  public void start() throws StageException {
+    for (String fileExtensionSupported : SUPPORTED_FILE_TYPES) {
```

Comment:
can this logic for populating a filehandlers array be extracted into a utility method?
we have to perform the same logic in BaseStorageClient where we have an initializeFileHandlers() method
I'm thinking, just like we have a static FileHandler.create() could we also have something like FileHandler.createFromConfig... returning a HashMap populated with all the handlers specified in the config?